validphys2/src/validphys/sumrules.py

```diff
@@ -123,24 +123,23 @@ def f(x, lpdf, irep, Q):
 }
 
 
+def _integral(rule_f, pdf_member, Q, config=None):
+    """Integrate `rule_f` for a given `pdf_member` at a given energy
+    separating the regions of integration. Uses quad.
+    """
+    if config is None:
+        config = {"limit":1000, "epsabs": 1e-4, "epsrel": 1e-4}
+    res = 0.0
+    lims = [(1e-9, 1e-5), (1e-5, 1e-3), (1e-3, 1)]
+    for lim in lims:
+        res += quad(rule_f, *lim, args=(pdf_member, Q), **config)[0]
+    return res
+
+
 def _sum_rules(rules_dict, lpdf, Q):
     """Compute a SumRulesGrid from the loaded PDF, at Q"""
-    nmembers = lpdf.n_members
-    #TODO: Write this in something fast
-    #If nothing else, at least allocate and store the result contiguously
-    res = np.zeros((len(rules_dict), nmembers))
-    integrands = rules_dict.values()
-    def integral(f, a, b, irep):
-        #We increase the limit to capture the log scale fluctuations
-        return integrate.quad(f, a, b, args=(lpdf, irep, Q),
-               limit=1000,
-               epsabs=1e-4, epsrel=1e-4)[0]
-
-    for irep in range(nmembers):
-        for r, f in enumerate(integrands):
-            res[r,irep] =  (integral(f, 1e-9, 1e-5, irep)  +
-               integral(f, 1e-5, 1e-3, irep) + integral(f, 1e-3, 1, irep))
-    return {k: v for k,v in zip(rules_dict, res)}
+    return {k: [_integral(r, m, Q) for m in lpdf.members] for k,r in rules_dict.items()}
+    # I very much doubt that looping over PDF members is the slow piece here...
```

Comment:
Sure, the loop is not making it slow, but why does that need to be mentioned?

Reply:
I guess I was responding to the previous comment :laughing: